File: mobile/apps/photos/plugins/ente_photos_platform/android/src/main/kotlin/io/ente/photos/platform/flutter/CountryNamesChannelAdapter.kt
package io.ente.photos.platform.flutter

import io.ente.photos.platform.country.CountryNamesService
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

internal class CountryNamesChannelAdapter : MethodChannel.MethodCallHandler {
    private val service = CountryNamesService()
    private lateinit var channel: MethodChannel

    fun attach(binding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(binding.binaryMessenger, CHANNEL)
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) =
        when (call.method) {
            "get" -> {
                val locale = call.arguments as? String
                    ?: return result.error("invalid_locale", "Expected a locale identifier", null)
                val names = service.names(locale)
                result.success(mapOf("region" to names.region, "names" to names.names))
            }

            else -> result.notImplemented()
        }

    fun detach() = channel.setMethodCallHandler(null)

    private companion object {
        const val CHANNEL = "io.ente.photos.platform/country_names"
    }
}

File: mobile/apps/photos/plugins/ente_photos_platform/ios/Flutter/CountryNamesChannelAdapter.swift
@preconcurrency import Flutter

@MainActor
final class CountryNamesChannelAdapter {
    private let channel: FlutterMethodChannel
    private let service = CountryNamesService()

    init(registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(
            name: "io.ente.photos.platform/country_names",
            binaryMessenger: registrar.messenger()
        )
        channel.setMethodCallHandler { [service] call, result in
            guard call.method == "get" else {
                result(FlutterMethodNotImplemented)
                return
            }
            guard let arguments = call.arguments as? [String: Any] else {
                result(FlutterError(code: "invalid_arguments", message: "Expected country name arguments", details: nil))
                return
            }
            guard let locale = arguments["locale"] as? String else {
                result(FlutterError(code: "invalid_locale", message: "Expected a locale identifier", details: nil))
                return
            }
            guard let nativeLocales = arguments["nativeLocales"] as? [String: [String]] else {
                result(FlutterError(code: "invalid_native_locales", message: "Expected native country locales", details: nil))
                return
            }
            let names = service.names(localeIdentifier: locale, nativeLocales: nativeLocales)
            result(["region": names.region, "names": names.names, "nativeNames": names.nativeNames])
        }
    }

    func detach() {
        channel.setMethodCallHandler(nil)
    }
}
